### DISCUSSION TITLE
The current fee design is not robust against spam attacks

### DISCUSSION THREAD

**Discussion by nymkappa** [icon-project/btp#64] [Ideas]
Hi.

I would like to start a discussion regarding the fee design in BTP, regarding the fee charged for making a BTP request.

### Current situation
Right now BTP native coin and token transfers charges a fee based on a ratio. For example:
* I transfer `10 ICX` from ICON to the destination chain
* I will receive `10 wrapped ICX * (1.0 - fee_ratio)` on the destination chain

The `fee_ratio` is set when uploading BSH contracts, but can also be changed through `setFeeRatio` smart contract function.

### The problem
The main issue with this design is that it makes it very easy to spam different networks with BTP messages. On ICON, transactions fee are very cheap. Most of the time they are under `0.01 ICX` (which is around `0.015 USD` at the time of writing this).

If someone wanted to run a DDos attack, it would require minimal amount of money and effort.

### Proposed solution to explore
We need to add a `fee_min` variable in BSH contracts. It is not very clear how this `fee_min` variable would be managed but I believe a good first approach would be to set this `fee_min` manually at first.

For example, the day we launch BTP, we could simply decide to charge a minimum of `0.50 USD` per BTP message. Using the current exchange rate of each network native coin, we would set the `fee_min` appropriately.

### Long term solution
Later on, maybe for BTP 2.0, we could have this `fee_min` set dynamically. Relays could be responsible for fetching the price of their respective networks and update the `fee_min` value. Another less centralized approach would be to use Band Protocol to compute `fee_min` directly in the BSH contract.

I hope this summary was clear. Please let me know what you think about this, and don't hesitate to point out flaws in that thinking. Also please share any other suggestion for protecting BTP network against spam.

Thank you.

**Comment by quocle108** [comment]
@jorisvial Yes, this looks good idea. 

Can you check this case too: There is another issue with current BTP that make it easy to spam. It nows return the full user's fund in case the failed transfer

**Reply by flowersinthesand** [reply]
I remember ICON team decided not to charge the BTP fee to avoid bad UX even if the transaction fails at first so we designed and implemented in a way that the fee is returned to the user in such a case but why don't we charge the BTP fee whether the transaction succeeds or not to defend the attacks?

The normal users may make some mistake and lose the fee if we choose this way. However, as they are likely to use services like the BTP dashboard, I think such services can protect them from making such mistakes in advance by verifying the destination address.

**Reply by nymkappa** [reply]
Yes I though about this too. If someone manages to send many BTP messages which are expected to fail on the destination chain, they could empty BMR's wallets.

I believe the min_fee should never been refunded to protect against any kind of spam attack.

**Comment by MuhammedIrfan** [comment]
@jorisvial As you proposed we should implement fee_min variable. Instead of having it in BSH would it be feasible to have it in BMC  and let BSH fetch fee_min from BMC? 
In future, as you proposed we can handle the price using band protocol in BMC for each registered network.

To add more to the context, Workflow would be

1. ICX holders decide the fee_min which we can get through voting.
2. Once approved ICON BMC would send BTP Messages to connected BMCs
3. Connected BMCs will update the fee_min.
4. BSH should fetch fee_min and convert with regards to the token/coin to be transferred using an API in BMC using Band protocol. *
5. Probably Fee should be a separate field in BTP Message. *
6. Validate if fees are min fees using band protocol in BMC. *
7. Proceed to send the message.

*Needs a little more brainstorming

**Reply by nymkappa** [reply]
It's better to keep all fee related variables at the same place. Right now the fee rate is set and stored in the BSH so for a first implementation it's just simpler to add it in the BSH.

But for a more advanced process like you've described we can redefine how we manage fee. It would maybe even make sense to have a separate contract where all fee for each coins are stored, and anyone/any contract can read the fees value.

**Reply by MuhammedIrfan** [reply]
sure jorisvial. Kindly help me clear the below doubts.

1. How do we set fee_min for Token BSH, Should we charge the fee in the native coin of the network.
2. How do we tackle volatility for the coin/tokens.

**Reply by nymkappa** [reply]
1. The fee_min will be set by an external call to the contract, only contract owner/operator can execute. I think it makes sense to charge the fee using the native coin of the network for simplicity. All we need to do is to make service function call payable and check if the paid amount is >= min_fee.
2. For now it should not be a real concern. We just want to avoid spam.

For BTP 1.0, we should set the min_fee when we deploy BTP contracts. For example, we decide that for the release, we want to charge a minimum fee of $0.50, we will compute the fx rate at the moment of deploying the contract and set the value. This will block spam abuse directly. We can write a basic script that update the min_fee once a day if we deem it necessary. But for now this would do the job I think.

Once BTP is running, we can use an oracle to set the min_fee dynamically. And then, the next step would be to have governance. But that's more like BTP 2.0 than BTP 1.0 in my opinion.

**Comment by nymkappa** [comment]
As @MuhammedIrfan pointed out, `min_fee` is a confusing notation. It should be named `fixed_fee` instead. It would be used the following way:

Fee paid = `fixed_fee_in_native_currency + fee_ratio * amount_of_token_transfered`

This is easy to implement, but makes the UX less intuitive, which I believe can be abstracted with BTP Dashboard.

**Comment by MuhammedIrfan** [comment]
Hi Team,

My suggestion is to have the fixed fee in the BMC/Separate Contract registered in BMC and let the BSHs fetch the fixed fee from the BMC. ICON network would be managing it through BTP message. Initially, it can be done as a manual process and in the future, we can use ICON governance to set the fees in terms of USD and get the help of an oracle to set the fees with regards to native network currency.

We will need to decide what goes into BTP 1.0. Also, we need to design the fixed fee considering the scenario
source network <==> route network <==> destination network.

Thanks
